File: 03-dynamic_programming/tight.cpp
#include "../header.hpp"
#include <iomanip>
#include <sstream>

ll f(ll prev, ll k, ll n, mi& cache)
{
	if (cache[prev][n] != -1) return cache[prev][n];
	if (n == 1) {
		cache[prev][n] = 1;
		return 1;
	}

	ll sum = 0;

	if (prev+1 <= k) 	sum += f(prev+1, k, n-1, cache);
	if (0 <= prev <= k) sum += f(prev,   k, n-1, cache);
	if (prev-1 >= 0) 	sum += f(prev-1, k, n-1, cache);
	
	cache[prev][n] = sum;
	return sum;
}

int main() {
	ll k_max = 9;
	ll n_max = 100;

	ll k, n;
	while (cin >> k >> n)
	{
		mi cache = mi(k_max+1, vi(n_max+1, -1));

		ll count = 0;
		for (ll i = 0; i <= k; i++)
			count += f(i, k, n, cache);

		lf ratio = (count / pow((k+1), n)) * 100;

		stringstream stream;
		stream << fixed << setprecision(9) << ratio;

		cout << stream.str() << endl;

		// cout << k << " " << n << " " << count << " / " << pow((k+1), n) << " = " << ratio << endl;

		// cout << "cache" << endl;
		// for (ll i = 0; i <= k; i++)
		// {
		// 	for (ll j = 0; j <= n; j++)
		// 		cout << cache[i][j] << "\t";
		// 	cout << endl;
		// }
	}
	
	return 0;
}

File: 03-dynamic_programming/tight.py
import sys


def f(prev, k, n, cache):
	if (cache[prev][n] != -1):
		return cache[prev][n]

	if (n == 1):
		cache[prev][n] = 1
		return 1

	s = 0

	if (prev+1 <= k): 		s += f(prev+1, k, n-1, cache)
	if (0 <= prev <= k): 	s += f(prev,   k, n-1, cache)
	if (prev-1 >= 0): 		s += f(prev-1, k, n-1, cache)
	
	cache[prev][n] = s
	return s


k_max = 9
n_max = 100

for line in sys.stdin:
	k, n = map(int, line.split())

	cache = [[-1]*(n_max+1) for _ in range(k_max+1)]

	count = 0
	for i in range(k+1):
		count += f(i, k, n, cache)

	ratio = (count / (k+1) ** n) * 100

	print(f'{ratio:.8f}')
